ros2/src/autorccar_costmap/src/costmap.cpp

```diff
@@ -0,0 +1,227 @@
+    cnt_limit = cfg["global"]["updateEveryNthLidar"].as<unsigned int>();
+
+    global_width = cfg["global"]["width"].as<int>();
+    global_height = cfg["global"]["height"].as<int>();
+
+    local_width = cfg["local"]["width"].as<int>();
+    local_height = cfg["local"]["height"].as<int>();
+
+    std::cout << "Successfully loaded the config" << std::endl;
+}
+
+void Costmap::InitCostmap(Eigen::MatrixXd& costmap, int size_x, int size_y) {
+    costmap.resize(size_x, size_y);
+    costmap.setConstant(ProbabilityToLogOdds(P_UNKNOWN));
+}
+
+void Costmap::UpdatePointCloud(const pcl::PointCloud<pcl::PointXYZI>::Ptr pcl) {
+    cur_point_cloud->clear();
+    cur_point_cloud_transformed->clear();
+    cur_point_cloud_filtered->clear();
+
+    pcl::copyPointCloud(*pcl, *cur_point_cloud);
+
+    UpdateCostmapFlag();
+}
+
+void Costmap::UpdatePose(Eigen::Matrix<double, 4, 4>& trans) { cur_pose = trans; }
+
+void Costmap::UpdateCostmapFlag() {
+    cnt_iter += 1;
+
+    if (cnt_iter >= cnt_limit) {
+        costmap_flag = true;
+        cnt_iter = 0;
+    }
+}
+
+void Costmap::UpdateCostmap() {
+    // Get cell indices of current pose. This will be used as the start point of the bresenham line.
+    int sx = floor(cur_pose(0, 3) / resolution) + global_center_x;  // assume (global_center_x, global_center_y) = (0,0)
+    int sy = floor(cur_pose(1, 3) / resolution) + global_center_y;
+
+    // Transform point cloud from body to world frame.
+    pcl::transformPointCloud(*cur_point_cloud, *cur_point_cloud_transformed, cur_pose);
+
+    // Filtering
+    pcl::PassThrough<pcl::PointXYZI> zFilter;
+    zFilter.setInputCloud(cur_point_cloud_transformed);
+    zFilter.setFilterFieldName("z");
+    zFilter.setFilterLimits(0.0, 1.0);
+    zFilter.filter(*cur_point_cloud_filtered);
+
+    for (const auto& pt : *cur_point_cloud_filtered) {
+        // Get cell indices of the pointcloud
+        int x = floor(pt.x / resolution) + global_center_x;
+        int y = floor(pt.y / resolution) + global_center_y;
+
+        // The cell where the lidar point located is occupied.
+        UpdateCell(x, y, P_OCCUPIED);
+
+        // The cells on the bresenham line are free.
+        BresenhamLine(sx, sy, x, y);
+    }
+
+    costmap_flag = false;
+}
+
+void Costmap::BresenhamLine(int sx, int sy, int ex, int ey) {
+    int dx = std::abs(ex - sx);
+    int dy = std::abs(ey - sy);
+
+    // discriminant
+    int d;
+
+    // current point
+    int x = sx;
+    int y = sy;
+
+    std::vector<std::pair<int, int>> line;
+
+    if (dx > dy) {
+        // when 0 < |slope| < 1
+        d = 2 * dy - dx;
+        for (int i = 0; i < dx; i++) {
+            if (d < 0) {
+                d = d + 2 * dy;
+            } else {
+                if (ey > sy) {
+                    y = y + 1;
+                } else {
+                    y = y - 1;
+                }
+                d = d + 2 * (dy - dx);
+            }
+            line.push_back(std::make_pair(x, y));
```

Comment:
```suggestion
            line.emplace_back({x, y});
```

push_back -> 임시객체 생성 후 복사
emplace_back -> 생성자 파라미터를 직접 넘겨서 vector 내부에서 생성자 호출

그리고 x, y 를 pair를 넣는다면
`std::vector<Eigen::Vec2i>` 가 편하지않을까요?!